### PR TITLE
Refactor text to use virtualfile_from_data

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

- Refactor text to use `virtualfile_from_data`
- Raise exceptions for invalid inputs
- Add more tests for invalid inputs

Related to #949.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/text.py [hunk 1]
```diff
@@ -157,15 +156,17 @@ def text_(
     # Ensure inputs are either textfiles, x/y/text, or position/text
     if position is None:
         kind = data_kind(textfiles, x, y, text)
```
Comment (meghanrjones) [inline pygmt/src/text.py:158]: It it worth protecting against passing both textfiles and x/y/text to data_kind()? Just affects what error the user gets.
```suggestion
        if (x is not None or y is not None) and textfiles is not None:
            raise GMTInvalidInput( "Provide either position only, or x/y pairs, or textfiles.")
        kind = data_kind(textfiles, x, y, text)
```

File: pygmt/src/text.py [hunk 2]
```diff
@@ -157,15 +156,17 @@ def text_(
...
-                "Provide either position only, or x/y pairs, not both"
+                "Provide either position only, or x/y pairs, or textfiles."
             )
-        kind = "vectors"
-
-    if kind == "vectors" and text is None:
-        raise GMTInvalidInput("Must provide text with x/y pairs or position")
+        if text is None or is_nonstr_iter(text):
+            raise GMTInvalidInput("Text can't be None or array.")
+        kind = "file"
```
Comment (meghanrjones) [inline pygmt/src/text.py:168]: It doesn't really matter, but "file" does not seem like an accurate descriptor here to me.
```suggestion
        kind = ""
```

**Review by meghanrjones — APPROVED** [review]
Looks good to me. I only had a couple small comments.

**Comment by seisman** [comment]
Thanks for your suggestions. I'll mark this PR as draft, because I don't like the current implemention.

File: pygmt/src/text.py [hunk 3]
```diff
@@ -203,19 +204,15 @@ def text_(
         extra_arrays.append(kwargs["t"])
         kwargs["t"] = ""
 
+    # Append text as the last column
+    # text must be in str type, see issue #706
```
[the remark this answers is not in the document]
Comment (seisman) [reply pygmt/src/text.py:224]: > text must be in str type

This statement is still True.

File: pygmt/src/text.py [hunk 4]
```diff
@@ -219,18 +220,14 @@ def text_(
...
```
Comment (seisman) [inline pygmt/src/text.py:224]:
```suggestion
    # Append text as the last column. Text must be passed in str type.
```

File: pygmt/src/text.py [hunk 5]
```diff
@@ -219,18 +221,13 @@ def text_(
         extra_arrays.append(kwargs["t"])
         kwargs["t"] = ""
 
+    # Append text as the last column. Text must be passed in str type.
```
Comment (weiji14) [inline pygmt/src/text.py:224]:
```suggestion
    # Append text at last column. Text must be passed in as str type.
```

File: pygmt/src/text.py [hunk 6]
```diff
@@ -164,24 +163,27 @@ def text_(
...
+                "Provide either position only, or x/y pairs, or textfiles."
             )
-        kind = "vectors"
-
-    if kind == "vectors" and text is None:
-        raise GMTInvalidInput("Must provide text with x/y pairs or position")
+        if text is None or is_nonstr_iter(text):
+            raise GMTInvalidInput("Text can't be None or array.")
+        kind = ""
+        textfiles = ""
```
Comment (weiji14) [inline pygmt/src/text.py:186]: Why are the `kind` and `textfiles` variables set to blank strings here?

**Review by weiji14 — COMMENTED** [review]
Had a quick glance and can't see any major issues with this refactor. Is there anything else remaining you want to work on here @seisman?

**Comment by seisman** [comment]
> Is there anything else remaining you want to work on here?

I plan to not make more reactors in this PR to make it easier to review.

File: pygmt/src/text.py (hunk 6 above)
Comment (weiji14) [reply pygmt/src/text.py:186]: Ok with `textfiles = ""`, didn't know about the `+tSTRING` syntax!

Looking at the `kind` variable, it seems to be an internal PyGMT-only variable that is set to bypass the if-condition `if kind == "vectors":` at L225. Maybe use `kind=None` instead?

```suggestion
        kind = None
        textfiles = ""
```

**Comment by meghanrjones** [comment]
Sorry for not getting back to this in time to be helpful

**Comment by seisman** [comment]
No worries. You can still leave your comments if there are any :smile:

**Pull request merged**